File: backend/app/models/team_builder.py
"""Durable drafts and provisioning state for intelligent team creation."""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import CheckConstraint, DateTime, ForeignKey, Index, Integer, String, Text, UniqueConstraint, func, text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class TeamBuildDraft(Base):
    """A user-reviewable plan. Draft generation must not create product resources."""

    __tablename__ = "team_build_drafts"
    __table_args__ = (
        CheckConstraint(
            "status IN ('generating', 'ready', 'invalid', 'confirmed', 'expired', 'cancelled')",
            name="ck_team_build_drafts_status",
        ),
        Index("ix_team_build_drafts_tenant_creator_updated", "tenant_id", "creator_user_id", "updated_at"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="RESTRICT"), nullable=False
    )
    creator_user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="RESTRICT"), nullable=False
    )
    requirement: Mapped[str] = mapped_column(Text, nullable=False)
    constraints: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict, server_default=text("'{}'::jsonb"))
    generated_plan: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    reviewed_plan: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    plan_version: Mapped[int] = mapped_column(Integer, nullable=False, default=1, server_default=text("1"))
    confirmed_plan_version: Mapped[int | None] = mapped_column(Integer, nullable=True)
    status: Mapped[str] = mapped_column(
        String(24), nullable=False, default="generating", server_default=text("'generating'")
    )
    error_code: Mapped[str | None] = mapped_column(String(100), nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    expires_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    confirmed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now()
    )


class TeamProvisionJob(Base):
    """Replay-safe asynchronous materialization of one confirmed draft."""

    __tablename__ = "team_provision_jobs"
    __table_args__ = (
        CheckConstraint(
            "status IN ('queued', 'validating', 'provisioning_agents', 'waiting_for_agents', "
            "'creating_group', 'activating', 'completed', 'retryable_failed', 'failed')",
            name="ck_team_provision_jobs_status",
        ),
        UniqueConstraint("draft_id", "idempotency_key", name="uq_team_provision_jobs_draft_idempotency"),
        Index("ix_team_provision_jobs_status_updated", "status", "updated_at"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="RESTRICT"), nullable=False
    )
    draft_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("team_build_drafts.id", ondelete="RESTRICT"), nullable=False
    )
    requesting_user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="RESTRICT"), nullable=False
    )
    idempotency_key: Mapped[str] = mapped_column(String(120), nullable=False)
    status: Mapped[str] = mapped_column(String(32), nullable=False, default="queued", server_default=text("'queued'"))
    error_code: Mapped[str | None] = mapped_column(String(100), nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    group_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("groups.id", ondelete="RESTRICT"))
    leader_participant_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("participants.id", ondelete="RESTRICT")
    )
    session_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("chat_sessions.id", ondelete="RESTRICT")
    )
    activation_message_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("chat_messages.id", ondelete="RESTRICT")
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now()
    )


class TeamProvisionMember(Base):
    """One planned roster item and its durable reuse/create outcome."""

    __tablename__ = "team_provision_members"
    __table_args__ = (
        CheckConstraint("source IN ('existing', 'new')", name="ck_team_provision_members_source"),
        CheckConstraint(
            "status IN ('pending', 'resolving', 'creating', 'waiting', 'ready', 'failed')",
            name="ck_team_provision_members_status",
        ),
        UniqueConstraint("job_id", "member_key", name="uq_team_provision_members_job_key"),
        Index("ix_team_provision_members_job_status", "job_id", "status"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("tenants.id", ondelete="RESTRICT"), nullable=False
    )
    job_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("team_provision_jobs.id", ondelete="CASCADE"), nullable=False
    )
    member_key: Mapped[str] = mapped_column(String(120), nullable=False)
    source: Mapped[str] = mapped_column(String(16), nullable=False)
    role_spec: Mapped[dict] = mapped_column(JSONB, nullable=False)
    status: Mapped[str] = mapped_column(String(16), nullable=False, default="pending", server_default=text("'pending'"))
    agent_id: Mapped[uuid.UUID | None] = mapped_column(UUID(as_uuid=True), ForeignKey("agents.id", ondelete="RESTRICT"))
    participant_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("participants.id", ondelete="RESTRICT")
    )
    error_code: Mapped[str | None] = mapped_column(String(100), nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now()
    )

File: frontend/src/types/teamBuilder.ts
/** Durable intelligent-team draft and provisioning contracts. */

export type TeamBuildDraftStatus =
    | 'generating' | 'ready' | 'invalid' | 'confirmed' | 'expired' | 'cancelled';
export type TeamProvisionJobStatus =
    | 'queued' | 'validating' | 'provisioning_agents' | 'waiting_for_agents'
    | 'creating_group' | 'activating' | 'completed' | 'retryable_failed' | 'failed';

export type TeamWorkflowPreset = 'default' | 'agile' | 'product_research' | 'custom';

export interface TeamPlanMember {
    member_key: string;
    name: string;
    role_description: string;
    responsibility: string;
    source: 'existing' | 'new';
    existing_agent_id: string | null;
    template_id: string | null;
    skill_ids: string[];
    is_leader: boolean;
}

export interface TeamPlanDelegation {
    from_member_key: string;
    to_member_key: string;
    instruction: string;
}

export interface TeamPlanWorkflowStage {
    key: string;
    title: string;
    goal: string;
    requires_approval: boolean;
}

export interface TeamPlanWorkflow {
    preset: TeamWorkflowPreset;
    name: string;
    stages: TeamPlanWorkflowStage[];
}

export interface TeamPlan {
    group_name: string;
    goal: string;
    assumptions: string[];
    phases: string[];
    members: TeamPlanMember[];
    delegations: TeamPlanDelegation[];
    workflow?: TeamPlanWorkflow | null;
}

export interface TeamBuildDraft {
    id: string;
    status: TeamBuildDraftStatus;
    requirement: string;
    constraints: Record<string, unknown>;
    generated_plan: TeamPlan | null;
    reviewed_plan: TeamPlan | null;
    plan_version: number;
    confirmed_plan_version: number | null;
    error_code: string | null;
    error_message: string | null;
    created_at: string;
    updated_at: string;
}

export interface TeamProvisionMember {
    member_key: string;
    source: 'existing' | 'new';
    status: 'pending' | 'resolving' | 'creating' | 'waiting' | 'ready' | 'failed';
    agent_id: string | null;
    participant_id: string | null;
    error_code: string | null;
    error_message: string | null;
}

export interface TeamProvisionJob {
    id: string;
    draft_id: string;
    status: TeamProvisionJobStatus;
    group_id: string | null;
    leader_participant_id: string | null;
    session_id: string | null;
    activation_message_id: string | null;
    error_code: string | null;
    error_message: string | null;
    members: TeamProvisionMember[];
}

export interface TeamProvisionJobSummary {
    id: string;
    draft_id: string;
    status: TeamProvisionJobStatus;
    group_id: string | null;
    session_id: string | null;
    error_message: string | null;
}

export interface TeamBuildHistoryItem {
    draft: TeamBuildDraft;
    job: TeamProvisionJobSummary | null;
}
